src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs

```diff
@@ -0,0 +1,392 @@
+        /// This will return the best provider which is a superset of wszProvider,
+        /// or NULL if there is no upgrade available on the machine.
+        /// </summary>
+        /// <param name="dwProvType">provider type</param>
+        /// <param name="wszProvider">provider name</param>
+        /// <returns>Returns upgrade CSP name</returns>
+        public static string UpgradeDSS(int dwProvType, string wszProvider)
+        {
+            string wszUpgrade = null;
+            if (string.Equals(wszProvider, MS_DEF_DSS_DH_PROV, StringComparison.Ordinal))
+            {
+                SafeProvHandle safeProvHandle;
+                // If this is the base DSS/DH provider, see if we can use the enhanced provider instead.
+                if (S_OK == AcquireCryptContext(out safeProvHandle,
+                    null,
+                    MS_ENH_DSS_DH_PROV,
+                    dwProvType,
+                    (uint)CryptAcquireContextFlags.CRYPT_VERIFYCONTEXT))
+                {
+                    wszUpgrade = MS_ENH_DSS_DH_PROV;
+                }
+                safeProvHandle.Dispose();
+            }
+            return wszUpgrade;
+        }
+
+        /// <summary>
+        /// Helper for DsaCryptoServiceProvider.ImportParameters()
+        /// </summary>
+        internal static byte[] ToKeyBlob(this DSAParameters dsaParameters, int algId)
+        {
+            if (algId != CapiHelper.CALG_DSS_SIGN)
+                throw new PlatformNotSupportedException();
+
+            const int NTE_BAD_DATA = unchecked((int)CryptKeyError.NTE_BAD_DATA);
+
+            // Validate the DSA structure first
+            // P and Q are required. Q is a 160 bit divisor of P-1.
+            if (dsaParameters.P == null || dsaParameters.P.Length == 0 || dsaParameters.Q == null || dsaParameters.Q.Length != DSS_Q_LEN)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            // G is required. G is an element of Z_p
+            if (dsaParameters.G == null || dsaParameters.G.Length != dsaParameters.P.Length)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            // If J is present, it should be less than the size of P: J = (P-1) / Q
+            // This is only a sanity check. Not doing it here is not really an issue as CAPI will fail.
+            if (dsaParameters.J != null && dsaParameters.J.Length >= dsaParameters.P.Length)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            // Y is present for V3 DSA key blobs, Y = g^j mod P
+            if (dsaParameters.Y != null && dsaParameters.Y.Length != dsaParameters.P.Length)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            // The seed is always a 20 byte array
+            if (dsaParameters.Seed != null && dsaParameters.Seed.Length != 20)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            bool isPrivate = (dsaParameters.X != null && dsaParameters.X.Length != 0);
```

Comment:
Good catch; thanks for looking at the native code. I added the check for == DSS_Q_LEN if the length is greater than 0. I believe the is a x mod q, so that makes sense, even though xbitlen is still part of the structure